### PR TITLE
Add optimize for visual prompting to 2.0

### PR DESCRIPTION
**Pull request by sungchul1**
### Summary

This PR includes:
- [x] optimize
    - inference results (fine-tuning, #test=150)
      |  | torch (F32, GPU) | torch (F32, CPU) | OVIR (F32, CPU) | OVIR (I8, CPU) |
      | -- | -- | -- | -- | -- |
      | F1 | 91.30 | 91.30 | 91.40 | 87.99 |
      | e2e time | 34s | 99s | 69s | 53s |

    - inference results (zero-shot, #test=150)
      |  | torch (F32, GPU) | torch (F32, CPU) | OVIR (F32, CPU) | OVIR (I8, CPU) |
      | -- | -- | -- | -- | -- |
      | F1 | 67.19 | 67.19 | 67.72 | 55.53 |
      | e2e | 12s | 108s | 80s | 113s |
- [x] updates
    - [x] `export_params`
    - [x] `_inspect_overlapping_areas`

**TODO**
- demo scenario of `exportable_code` for visual prompting tasks should be decided.
- `segmentation fault` occurs after `optimize` with current openvino version (2023.3).

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py [hunk 1]
```diff
@@ -415,11 +415,20 @@ def _calculate_mask_iou(mask1: Tensor, mask2: Tensor) -> Tensor:
...
+                _mask_iou = _calculate_mask_iou(mask, other_mask)
+                if _mask_iou > threshold_iou:
+                    # compare overlapped regions between different labels and filter out the lower score
                     if used_points[label][im][2] > used_points[other_label][jm][2]:
                         overlapped_other_label.append(jm)
                     else:
                         overlapped_label.append(im)
+                elif _mask_iou > 0:
+                    # refine the slightly overlapping region
+                    overlapped_coords = torch.where(torch.logical_and(mask, other_mask))
```
Comment (vinnamkim) [inline src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: Reuse `torch.logical_and(mask, other_mask)` at L404 not to compute it twice

File: src/otx/core/model/entity/visual_prompting.py [hunk 2]
```diff
@@ -838,11 +953,19 @@ def _calculate_mask_iou(mask1: np.ndarray, mask2: np.ndarray) -> np.ndarray:
...
-                if _calculate_mask_iou(mask, other_mask) > threshold_iou:
+                _mask_iou = _calculate_mask_iou(mask, other_mask)
+                if _mask_iou > threshold_iou:
                     if used_points[label][im][2] > used_points[other_label][jm][2]:
                         overlapped_other_label.append(jm)
                     else:
                         overlapped_label.append(im)
+                elif _mask_iou > 0:
+                    # refine the slightly overlapping region
+                    overlapped_coords = np.where(np.logical_and(mask, other_mask))
```
Comment (vinnamkim) [inline src/otx/core/model/entity/visual_prompting.py]: Why does here `_inspect_overlapping_areas()` looks same as the one in src/otx/algo/visual_prompting/zero_shot_segment_anything.py?

File: src/otx/algo/visual_prompting/zero_shot_segment_anything.py (hunk 1 above)
Comment (sungchul1) [reply src/otx/algo/visual_prompting/zero_shot_segment_anything.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/3040/commits/d248f81cf9ac949984d986ad1730b388a57131cf.

File: src/otx/core/model/entity/visual_prompting.py (hunk 2 above)
Comment (sungchul1) [reply src/otx/core/model/entity/visual_prompting.py]: torch version and numpy version. It is needed to merge two implementations as you said, but it is out of scope of this PR. I'll touch it next time.

**Review by vinnamkim — DISMISSED** [review]

File: tests/integration/cli/test_export_inference.py [hunk 3]
```diff
@@ -230,15 +226,21 @@ def run_cli_test(test_recipe: str, checkpoint_path: str, work_dir: Path, device:
         exported_model_path,
     ]
 
-    run_main(command_cfg=command_cfg, open_subprocess=fxt_open_subprocess)
+    run_main(
+        command_cfg=command_cfg,
+        open_subprocess=False if task in ("visual_prompting", "zero_shot_visual_prompting") else fxt_open_subprocess,
```
Comment (sungchul1) [inline tests/integration/cli/test_export_inference.py]: @yunchu I temporarily added it not to open subprocess, is it fine?

Comment (vinnamkim) [reply tests/integration/cli/test_export_inference.py]: Please revert it and add xfail mark if needed as discussed in the Teams chat

Comment (sungchul1) [reply tests/integration/cli/test_export_inference.py]: Done at https://github.com/openvinotoolkit/training_extensions/pull/3040/commits/e3d29aecd9e0d68544197beb6793bf3f486ff96e.

**Review by sungmanc — APPROVED** [review]

File: src/otx/core/model/entity/visual_prompting.py [hunk 4]
```diff
@@ -57,13 +62,37 @@ def _export_parameters(self) -> dict[str, Any]:
         export_params = super()._export_parameters
         export_params["metadata"].update(
             {
-                ("model_info", "model_type"): "segment_anything",
+                ("model_info", "model_type"): "Visual_Prompting",
                 ("model_info", "task_type"): "visual_prompting",
             },
         )
         export_params["input_size"] = (1, 3, self.model.image_size, self.model.image_size)
+        export_params["resize_mode"] = "fit_to_window"
```
Comment (eunwoosh) [inline src/otx/core/model/entity/visual_prompting.py:70]: Doesn't 'Pad` exist after resize? If it exists, need to change it to 'fit_to_window_letterbox'

File: src/otx/core/model/entity/visual_prompting.py [hunk 5]
```diff
@@ -225,6 +256,90 @@ def _customize_outputs(
...
+
+            image = images[0]["images"]  # use only the first image
+            if module == "image_encoder":
+                # resize
+                resized_image = self.model["image_encoder"].resize(
+                    image[0],
+                    (self.model["image_encoder"].w, self.model["image_encoder"].h),
+                )
+
+                # pad image if necessary because `fit_to_window` resize for python in modelapi doesn't support pad
```
Comment (eunwoosh) [inline src/otx/core/model/entity/visual_prompting.py:288]: ModelAPI support resize + pad by 'fit_to_window_letterbox'. Honestly, I don't know what it is for exactly. But should this part be changed if ModelAPI can support pad?

File: src/otx/core/model/entity/visual_prompting.py (hunk 4 above)
Comment (sungchul1) [reply src/otx/core/model/entity/visual_prompting.py:70]: `fit_to_window` already does `pad` after `resize` like below.
https://github.com/openvinotoolkit/model_api/blob/41c20267122bed9fe9b21631166d15818f669379/model_api/python/model_api/adapters/utils.py#L358-L367
image

File: src/otx/core/model/entity/visual_prompting.py (hunk 5 above)
Comment (sungchul1) [reply src/otx/core/model/entity/visual_prompting.py:288]: I reported the issues. https://github.com/openvinotoolkit/model_api/issues/167
`fit_to_window` with numpy doesn't do pad after resize, but one with opset does pad after resize.
So, if we want to use numpy version, padding after reisze is required.

File: src/otx/core/model/entity/visual_prompting.py (hunk 4 above)
Comment (eunwoosh) [reply src/otx/core/model/entity/visual_prompting.py:70]: As far as I know, "fit_to_window" executes `resize_image_with_aspect_ocv` function https://github.com/openvinotoolkit/model_api/blob/41c20267122bed9fe9b21631166d15818f669379/model_api/python/model_api/adapters/utils.py#L476

Comment (sungchul1) [reply src/otx/core/model/entity/visual_prompting.py:70]: It is numpy version, but through `embed_preprocessing` in adapter, opset version will be used.

Comment (sungchul1) [reply src/otx/core/model/entity/visual_prompting.py:70]: But they do differently as I left a comment the below.

**Review by eunwoosh — APPROVED** [review]

**Review by vinnamkim — APPROVED** [review]
